app/src/main/java/org/schabi/newpipe/util/FilePickerActivityHelper.java

```diff
@@ -131,12 +142,44 @@ public void onClickOk(@NonNull final View view) {
             super.onClickOk(view);
         }
 
+        /**
+         * @param file File getting the extension from.
+         * @return The file extension. If file has no extension, it returns null.
+         */
+        @Nullable
+        private String getExtension(@NonNull final File file) {
+            final String path = file.getPath();
+            final int i = path.lastIndexOf(".");
+            if (i < 0) {
+                return null;
+            } else {
+                return path.substring(i);
+            }
+        }
+
         @Override
         protected boolean isItemVisible(@NonNull final File file) {
             if (file.isDirectory() && file.isHidden()) {
                 return true;
             }
-            return super.isItemVisible(file);
+            final boolean ret = super.isItemVisible(file);
+            if (ret && !isDir(file) && !filterFileExtension.isEmpty()
+                    && (mode == MODE_FILE || mode == MODE_FILE_AND_DIR)) {
+                final String ext = getExtension(file);
+                return ext != null && filterFileExtension.equalsIgnoreCase(ext);
+            }
+            return ret;
```

Comment:
Could you add unit tests to this part of the logic. It seems to not rely on any anroid framework stuff and should be testable if you extract it into a seperate method or even class.

Reply:
Sure, but I have to look into it. I'm not the best one for Unit testing (at least if it isn't basic method testing ^^). Should I add a normal file, for example an empty txt-file, to the resources folder (like you did with the zips) )and check whether it's listed then?

Btw, the snippet is a recipe from the official doc if you're interested: http://spacecowboy.github.io/NoNonsense-FilePicker/example/filter_file_extension/

Reply:
>Sure, but I have to look into it. I'm not the best one for Unit testing (at least if it isn't basic method testing ^^).

That's a big part of profesional development, so it's worth to look into it. Especially because "bad" code is usually not testable und thus by forcing tests the quality rises by itself.

In this case it could be a bit more annoying to test, because you would have to prepare different folders with specific files for your tests (I think). If you have problems i can help you there.

Reply:
professional as in code that needs to be maintained aka changed

Reply:
I'm aware of the importance and logic isn't hard - just writing tests combined with UI-related stuff isn't good to handle (or Android/Activity lifecycle etc.). As you can see above, the snippet has the `mode` check in it, so I'll have to lift it up since this part isn't testable in (instrumental) tests (is it?).

Reply:
`mode` is simply an `int` which you can pass into the method.
I also don't see any UI-related stuff in this method, it's just that the method is in the `fragment` class. But you can create a `manager` class which only contains logic. The method would looks something like this
```java 
public boolean aa(File file, int mode, boolean parent) {
    if (file.isDirectory() && file.isHidden()) {
        return true;
    }
    if (parent && !file.isDirectory() && !filterFileExtension.isEmpty()
            && (mode == MODE_FILE || mode == MODE_FILE_AND_DIR)) {
        final String ext = getExtension(file);
        return filterFileExtension.equalsIgnoreCase(ext);
    }
    return parent;
}
```
You would then also move your `getExtension()` and `filterFileExtension` into that class. Or maybe pass in `filterFileExtension` as parameter to the method. You could pass in `parent` as an lambda which you can then lazily evaluate if it is needed (kotlin it's possible, java maybe?)

and called from the fragment like 
```java
return new Manager().aa(file, mode, super.isItemVisible(file));
```

----

On a side note. I checked out your code and Android Studio says that the `ext != null` is unnecessary because `equalsIgnoreCase` already performs a null check. So please remove that.